File: C#/ToyLanguage/ToyLanguage/Model/Expressions/ReadHeap.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ToyLanguage.Interfaces;

namespace ToyLanguage.Model.Expressions
{
    class ReadHeap : IExpressions
    {

    /**
     * Variable name
     */
    private string variableName;
        
    /**
     * The constructor
     * @param variableName The variable name
     */
    public ReadHeap(string variableName)
    {
        this.variableName = variableName;
    }

        public int eval(IMyDictionary<string, int> table)
        {
            throw new NotImplementedException();
        }

        public int eval(IMyDictionary<string, int> table, IHeap<int, int> heap){
        return heap.lookUp(table.lookUp(variableName));
        }

        public string MyToString()
        {
            return "ReadHeap( " + variableName + " )";
        }

    }

}

File: Java/src/model/Expresions/ReadHeap.java
package model.Expresions;

import Exceptions.DivideByZeroException;
import Exceptions.ValueNotFoundException;
import interfaces.Expression;
import interfaces.IDictionary;
import interfaces.IHeap;

/**
 * Created by Lucian on 11/23/2015.
 */
public class ReadHeap implements Expression {

    /**
     * Variable name
     */
    private String variableName;

    /**
     * The constructor
     * @param variableName The variable name
     */
    public ReadHeap(String variableName) {
        this.variableName = variableName;
    }

    @Override
    public int eval(IDictionary<String, Integer> table, IHeap<Integer, Integer> heap) throws ValueNotFoundException, DivideByZeroException {
        return heap.lookUp(table.lookUp(variableName));
    }

    @Override
    public String toString() {
        return "ReadHeap( "+ variableName + " )";
    }
}
